### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by Anastas0v**

### CODE REVIEW

File: project/firstproject/pom.xml [hunk 1]
```diff
@@ -21,11 +21,53 @@
...
+		</dependency> -->
+		<dependency>
+			<groupId>org.webjars</groupId>
+			<artifactId>jquery</artifactId>
+			<version>3.4.1</version>
+		</dependency>
+		<dependency>
+			<groupId>org.webjars</groupId>
+			<artifactId>bootstrap</artifactId>
+			<version>4.3.1</version>
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/pom.xml]: Extract version number <version>4.3.1</version> to variable at the top of the file:
Example:
```
   <properties>
        <bootstrap.version>4.3.1</bootstrap.version>
   </properties>

```
Then use it in this way:
`<version>${bootstrap.version}</version>`

File: project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import com.example.firstproject.services.AuthenticationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+public class RegisterController
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: You should add initial route for this Controller after @Controller annotation:
Example:
`@RequestMapping(value = "/auth")`

File: project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        Matcher matcher = pattern.matcher(email);
+        if (!matcher.matches()) {
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean checkUserName(String userName) {
+        if(userName.isEmpty() || userName.length() <= 2)
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java]: null check missing

**Review by aleksandar-bujaroski-armedia — CHANGES_REQUESTED** [review]

File: project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping(value = "/auth")
+public class RegisterController
+{
+    @Autowired
+    private AuthenticationService authenticationService;
+
+    @GetMapping("/register/registerPage")
```
Comment (dejannastevski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: Minor: change this to register-page

File: project/firstproject/pom.xml [hunk 5]
```diff
@@ -15,17 +15,71 @@
...
+			<scope>runtime</scope>
+		</dependency>
+<!--		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-security</artifactId>
+		</dependency> -->
+		<dependency>
+			<groupId>org.webjars</groupId>
+			<artifactId>jquery</artifactId>
+			<version>3.4.1</version>
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/pom.xml]: You should extract this version to variable

File: project/firstproject/pom.xml [hunk 6]
```diff
@@ -15,17 +15,71 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.webjars</groupId>
+			<artifactId>bootstrap</artifactId>
+			<version>${bootstrap.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-crypto</artifactId>
+			<version>5.6.1</version>
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/pom.xml]: This one too

File: project/firstproject/src/main/java/com/example/firstproject/models/User.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
+package com.example.firstproject.models;
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/models/User.java]: Package name should be singular -> **models** vs **model**

File: project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationService.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.example.firstproject.services;
+
+import com.example.firstproject.models.User;
+
+public interface AuthenticationService
+{
+    boolean checkEmail(String email);
+    boolean checkUserName(String userName);
+    boolean checkPassword(String password);
+    String validate(User user);
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationService.java]: Change method name to be more descriptive.
Example: validateUserRegistrationInfo

File: project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java]: Fix code formatting

File: project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public void saveUser(User user) {
+        userRepository.save(user);
+    }
+
+    @Override
+    public User encryptPassword(User user) {
+        User newUser = new User();
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/services/AuthenticationServiceImpl.java:65]: This create whole new user object, change method name to reflect method logic

File: project/firstproject/src/main/java/com/example/firstproject/util/PasswordHashing.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PasswordHashing {
+    @Bean
+    public PasswordEncoder encoder() {
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/util/PasswordHashing.java]: Change code formatting

File: project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java (hunk 2 above)
Comment (aleksandar-bujaroski-armedia) [reply project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: Change controller name to **AuthController** if you'll use this controller for login too

**Review by aleksandar-bujaroski-armedia — CHANGES_REQUESTED** [review]

File: project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java (hunk 2 above)
Comment (Anastas0v) [reply project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: i want to make another controller for login

Comment (aleksandar-bujaroski-armedia) [reply project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: Then change the **@RequestMapping(value = "/auth")** to **@RequestMapping(value = "/register")** and remove **register** from route name

File: project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            return "login";
+        }
+        else
+        {
+            model.addAttribute("errorMessage", result);
+            return "index";
+        }
+    }
+
+    @GetMapping("/login")
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/controller/RegisterController.java]: Remove login if you create new controller

File: project/firstproject/src/main/java/com/example/firstproject/controller/AuthenticationController.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import com.example.firstproject.services.AuthenticationService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Controller
+@RequestMapping(value = "/auth")
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/controller/AuthenticationController.java:13]: Change the route to be **@RequestMapping(value = "/register")** if you create new controller for login and remove register from other routes names:
Example:  `@GetMapping("/register/register-page")` to ` @GetMapping("/register-page")`

File: project/firstproject/src/main/java/com/example/firstproject/controller/AuthenticationController.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            return "login";
+        }
+        else
+        {
+            model.addAttribute("errorMessage", result);
+            return "index";
+        }
+    }
+
+    @GetMapping("/login")
```
Comment (aleksandar-bujaroski-armedia) [inline project/firstproject/src/main/java/com/example/firstproject/controller/AuthenticationController.java:43]: Remove login if you create new login controller

**Review by aleksandar-bujaroski-armedia — CHANGES_REQUESTED** [review]

**Review by aleksandar-bujaroski-armedia — APPROVED** [review]

**Review by dejannastevski-armedia — APPROVED** [review]

**Pull request merged**